CMakeLists.txt

```diff
@@ -0,0 +1,86 @@
+cmake_minimum_required(VERSION 3.7)
+
+project(orocos_kdl_vendor)
+
+find_package(ament_cmake REQUIRED)
+
+option(FORCE_BUILD_VENDOR_PKG
+  "Build orocos_kdl from source, even if system-installed package is available"
+  OFF)
+
+if(NOT FORCE_BUILD_VENDOR_PKG)
+  find_package(orocos_kdl QUIET)
+  # TODO(jacobperron): A package config is not provided. How to detect if python bindings are installed?
```

Comment:
It's lower on my priority list, since we are not currently depending on on the python bindings. If I don't get to it before merging this PR, then I'll create a ticket.

Reply:
There's a lot of users who actually want the Python bindings here.  Porting of the python bindings for e.g. tf2_kdl has been blocked on this for a long time.

That said, I'm fine with starting without that functionality, and then enabling it in a follow-up pull request.

Reply:
Okay, I spent some time looking at vendoring the Python bindings.

I've broken it up into two vendor packages, `orocos_kdl_vendor` and `python_orocos_kdl_vendor`. The latter depends on the former. This was more straight-forward than trying to link one CMake external project on another and it's more flexible since users who don't care about Python support don't have to build/install it.

In order to get the recent upstream changes to skip building pybind11, I'm now pointing to a newer commit (dbee885).

I'm able to build both packages locally (if system packages for orocod_kdl are not installed), however I'm still not able to import PyKDL. I'm not sure if it's something that should be fixed in the vendor package or upstream; further investigation required.

Reply:
Okay, sothe problem is that the shared library for the Python bindings is not a part of PYTHONPATH after sourcing the workspace. It seems pykdl relies on catkin for the env hook (see https://github.com/orocos/orocos_kinematics_dynamics/pull/285). I can add an env hook to the vendor package, but I'm not sure how to determine the path of the library in the binary directory; [it is dependent on the Python version](https://github.com/orocos/orocos_kinematics_dynamics/blob/4f705fe9e1533802578b5275ddad02b065b2be4e/python_orocos_kdl/CMakeLists.txt#L28). There must be some cmake spell to get this done, but I guess we could also duplicate the logic from upstream, though it seems brittle. If `catkin` is installed, then the install path changes.

Reply:
I've determined how to get the location of the shared library in the binary directory easily, with `ExternalProject_Get_Property`. Next, I tried to install it with [ament_python_install_module](https://github.com/ament/ament_cmake/blob/199ac7a046cbf686912a198a41a501858382af60/ament_cmake_python/cmake/ament_python_install_module.cmake), which _almost_ works except that the file doesn't exist at the configure step.. so the build never happens. Here's what the Cmake looks like:

```cmake
  # Build PyKDL
  externalproject_add(pykdl
    URL https://github.com/orocos/orocos_kinematics_dynamics/archive/e25a13fc5820dbca6b23d10506407bca9bcdd25f.zip
    URL_MD5 8ac40c9f84eca10dbe1f3777ba715ba0
    TIMEOUT 600 
    SOURCE_SUBDIR python_orocos_kdl
    INSTALL_COMMAND ""
    CMAKE_ARGS
      ${extra_cmake_args}
  )

  # Install Python module
  ExternalProject_Get_Property(pykdl BINARY_DIR)
  if(WIN32)
    set(pykdl_module "${BINARY_DIR}/PyKDL.pyd")
  else()
    set(pykdl_module "${BINARY_DIR}/PyKDL.so")
  endif()

  ament_python_install_module(${pykdl_module}
```

:confused:

Reply:
Okay, I just copied the relevant parts from `ament_python_install_module` since we don't seem to be able to call it directly.

@clalancette @nuclearsandwich PTAL [556def6](https://github.com/ros2/orocos_kdl_vendor/pull/1/commits/556def6bdcb8b008fe45a22c1ed5ebb4c964b3ad)

Reply:
> I can add an env hook to the vendor package, but I'm not sure how to determine the path of the library in the binary directory; [it is dependent on the Python version](https://github.com/orocos/orocos_kinematics_dynamics/blob/4f705fe9e1533802578b5275ddad02b065b2be4e/python_orocos_kdl/CMakeLists.txt#L28). There must be some cmake spell to get this done, but I guess we could also duplicate the logic from upstream, though it seems brittle. If `catkin` is installed, then the install path changes.

One solution is to add optional ament_cmake handling to upstream directly so that it could manage its own env hooks on ROS 2 as well. I think it could co-exist with the catkin setup in package.xml through the use of conditional dependencies. If we did that, it would make it possible to release upstream directly into ROS 2, which may obviate the need for vendor packages entirely at the cost of removing our ability to work with the platform version of orocos_kdl.

Reply:
I'm okay with that option, though is there a concern about side-by-side installing the version of orocos_kdl shipped with ROS the upstream version?

Reply:
I noticed that orocos_kdl is not released in Noetic (I guess maybe we're depending on upstream in that case?).

@MatthijsBurgh Maybe maintainers of orocos_kdl have a preference? Would it be better for ROS 2 users to depend on the version of orocos_kdl shipped with Ubuntu or a version released into ROS 2 itself? The latter option requires someone make regular releases into ROS 2.

Reply:
> If we did that, it would make it possible to release upstream directly into ROS 2, which may obviate the need for vendor packages entirely at the cost of removing our ability to work with the platform version of orocos_kdl.

This contradicts the conclusions (and my own agreement with them) in https://github.com/ros2/ros2/pull/1208#issuecomment-1007533586

I'm trying to unpack my own attitude. I think it comes down to how much logic we want in the vendor package itself. The vendor package in its ideal form is just a wrapper around an upstream project which provides the glue to build it on platforms which don't currently provide it.

The conflict for me is between logicless vendor packages (just enough to get the vendored package built and bolted in) and keeping packages _out_ of ROS when we could get them as platform-provided dependencies.

If in the long view we're trying to keep orocos_kdl as a platform dependency and willing, as maintainers and as a community, to deal with the trade-offs of working with platforms development cycles then we should try to make the vendor packages work since hopefully someday they'll be gone when all our tier 1 & 2 platforms provide them.
If we're expecting to want to develop and maintain orocos_kdl together with ROS 2 the way we do with some rmw providers then the equation flips and we may as well embrace it as a native ROS package now in order to avoid dealing with this vendor logic. My current understanding is that we'd be very happy with orocos_kdl from upstream and thus we should try to figure out how to keep the vendor logic low, but also not give up on the vendor packages.

Reply:
> I'm not sure how to determine the path of the library in the binary directory; [it is dependent on the Python version](https://github.com/orocos/orocos_kinematics_dynamics/blob/4f705fe9e1533802578b5275ddad02b065b2be4e/python_orocos_kdl/CMakeLists.txt#L28)

Opened https://github.com/ros2/orocos_kdl_vendor/pull/2 to solve that problem using `FetchContent`, `add_subdirectory()`, and `get_directory_property()`.

Reply:
I recommend to only use 1.5.X versions. Ubuntu has only released a 1.5.X on Jammy (22.04). So I think it would be best to release one by ROS2.